### DISCUSSION TITLE
Freeze graph during execution?

### DISCUSSION THREAD

**Discussion by dfdx** [apache/airflow#26863] [General]
We noticed that an already running DAG is updated if we add new or delete old tasks. For example, in graph below we inserted a couple of new print tasks and removed the last one (`print10`), and the execution process followed the updated structure, not the one it was started with. 

![image](https://user-images.githubusercontent.com/428171/193863343-03f45692-e0ce-4298-8f7e-50a885754e67.png)

Is it the intended behavior? Is there a way to freeze a graph for a particular run?

**Comment by potiuk** [comment]
Not possible now. AIp-36 which is not yet finalized, discussed and approved should adress it in the future. https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-36+DAG+Versioning

**Reply by dfdx** [reply]
Thanks for the reply! As far as I can see, AIP-36 is about changing the UI behavior and not DAG execution itself. I think the part relevant to us is at Scenario 3 section: 

> Airflow handles this situation based on the latest version of the DAG. The DAG run continues with task B (from the previous version) being skipped and task C now being run.  
> There is a school of thought that an alternative behaviour of:
> * The first version of the DAG including all scheduled tasks i.e. tasks B and C from the first version should run to completion.
> * And the second version of the DAG should be independent of the first version. 
>
> Though this was considered, this is a significant change to the current run-time behaviour of Airflow and not in the scope of the stated goal of this proposal. 

I also found [this comment](https://stackoverflow.com/a/67326263) from the AIP's author:

> However, what we have decided is that we will accomplish [Remote DAG Fetcher](https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-5+Remote+DAG+Fetcher) + DAG Versioning and enable versioning of DAG on the worker side, so a user will be able to run a DAG with the previous version too.
> 
> Currently, we don't have a date but mostly planning to do it around the end of 2021.

Since it's almost a year overdue, I guess this change is of low priority and won't land anytime soon. 

Anyway, thank you for working on the great tool!

**Comment by potiuk** [comment]
AIP-36 is not yet finished and it should cover both DAG execution and UI. Those to are tied together in the way that it cannot be separated. There is no way to implemnent UI side of dag versioning without versioning of the execution (unless you have a good idea ho to do it - but IMHO it is just logically impossible). But feel free to change my mind if you think you can somehow separate those two. I personally think it is just impossble.

> Since it's almost a year overdue, I guess this change is of low priority and won't land anytime soon.

As evertything in open-source - it's done when someone does it. This basicaly means that if no-one actively works on it, it is klinda lower priority. But it's easy to change - anyone could spend time and effort to continue leading the discussion and leadint the AIP to approval and later to implementation. This is how it works in open-source. You can also be one of the people to do it (or maybe even lead it to completion). The important thing here is that this is not a commercial project that you might want to "expect" something (mainly because you are not paying anything for it) - but if you think priority is important, investing your time and effort to move it forward (if no-one else is doing it) is a great idea. Feel free to increase the priority - by investing your effort to do so.

**Reply by dfdx** [reply]
To clarify, my previous comment is just a summary for a random reader like myself - the AIP-36 has a bit outdated description (see the cited snippets) and nobody is actively working on it. I certainly do not expect the maintainers to care about my problem more then myself or take any additional actions. In fact, even I don't care about this specific problem enough to put it to the front of my own todo list (including open source contributions). So things just go the way they should. 

Once again, thanks for your contributions!
